### ISSUE TITLE
Can't Select Worldspace in Worldspace Browser

### ISSUE THREAD

**Issue by AndalayBay** [TES5Edit/TES5Edit#1054]
**What is the version of xEdit you are using?**
TES4Edit 4.0.4 x64 (20781F15)

**Describe the bug**
Open Worldspace Browser via Apply Script... or Alt+F3

**To reproduce**
Select Wordspace [sic], then Select or press F3. Only the Oblivion worldspaces are listed. I have generated LOD for our project, so there are LOD textures for the worldspace but it's not listed.

**Expected behavior**
Our worldspace to be listed for selection. Note that our worldspace is in our master plugin, not in Oblivion.

Minor bug - the menu item is "Wordspace" instead of "Worldspace". ;)

I have been looking at the worldspace browser script, but I wasn't able to get the RAD Studio development environment running. It threw tons of errors. I wound up installing the Pascal language module for Visual Studio and now I don't get any errors, but I can't debug it. Anyway if someone can tell me how to fix the script, I can give that a try. I'm pretty sure it's the HasLOD function that's failing. Or the script is hard-coded to only check Oblivion.esm. I didn't see anything like that though.

I was going to ask on your Discord channel, but I DO NOT HAVE A CELL PHONE. Yes, you read that right. :D

**Comment by jbostrus** [comment]
I'll admit I'm not the most versed in that script. But it looks like the HasLod function is specifically looking for the LOD texture for 0,0 with the LOD level of 32 ***inside*** loaded archives. If the file is loose it will fail and thus report no LOD. The particular parts of relevant code are lines 89, 91, 119.
```
89: if not aNormals then texname := 'textures\landscapelod\generated\%s.%.2d.%.2d.%d.dds'

91: Result := Format(texname, [IntToStr(aFormID and $FFFFFF), x, y, aLODLevel]);

119: Result := ResourceExists(LODTextureFileName(wbGameMode, FormID(wrld), EditorID(wrld), 0, 0, 32, False))
```
So if you haven't packed the LOD to the plugin's related archive, do that.

Untested, but a possible alternative is to replace line 119 with an **or**
```
Result := ResourceExists(LODTextureFileName(wbGameMode, FormID(wrld), EditorID(wrld), 0, 0, 32, False)) or FileExists(DataPath + LODTextureFileName(wbGameMode, FormID(wrld), EditorID(wrld), 0, 0, 32, False))
```

**Comment by AndalayBay** [comment]
I don't think it's looking for the LOD to be in an archive. I'm pretty sure it's looking for the loose files to be in textures\LandscapeLOD\Generated. We did some testing and discovered that if there are no files in that directory, the script will fail with an unhandled exception and you have to use task manager to shut xEdit down (another bug if someone wants to fix it). It works fine when I generate the LOD and there are files in that directory.

The files are named dd.xx.yy.32.dds and dd.xx.yy.32_fn.dds for Oblivion where dd is the decimal version of the form ID for the worldspace. Tamriel is 0000003C which is 60 in decimal. xx and yy are the quad coordinates. So the HasLOD function is looking for 60.00.00.32.dds or 60.00.00.32_fn.dds.

Now we haven't used TES4Gecko to move our worldspaces yet. I wonder if that's the issue... Although I am only loading our ESM so the form ID for our worldspace starts with 01, so I thought it would work. I have verified that the names of the LOD files are the decimal value for the form ID for our worldspace.

Having said all that, I could certainly try packing the files into an archive.

**Comment by zilav** [comment]
It looks for a LOD texture at 0,0 quad, either loose or packed. However worldspaces in Oblivion work correctly only for 0 load index (unless some OBSE plugin fixed it) and the script accounts for that `aFormID and $FFFFFF`. So yes, you need to move worldspace or just remove `and $FFFFFF` from the script.

**Comment by AndalayBay** [comment]
Awesome! Thank you. Yes, we are using EngineBugFixes to get around the master load order bug. We need to see trees and other LOD in the CS so we can't move world spaces until we're done development. And absolute genius to whoever figured that out...

Thanks again, zilav.

**Comment by AndalayBay** [comment]
That allowed me to select our worldspace, but I got this error:
![sshot-5](https://user-images.githubusercontent.com/1171156/201463079-7e6c94f5-97ee-400e-96d6-c16829e114dc.png)
What does that mean? I didn't change any of the settings. If I need to, could you please tell me what I need to change?

**Comment by zilav** [comment]
Don't use x64 version of xEdit, use the ordinary 32 bits.

**Comment by AndalayBay** [comment]
All sorted - thanks. We ran into memory issues with the 32 bit version so had to switch to 64.

**Comment by AndalayBay** [comment]
Is there a way to save an overlay in the image? I can take a screenshot, but it would be nice to directly save the image with the overlay(s).

**Comment by zilav** [comment]
Replace with `imgOver` in `imgMap.Picture.Bitmap.SaveToFile(dlgSave.FileName);`

**Comment by AndalayBay** [comment]
Close, but not quite. It just saves the overlay, not the map itself. I can save them as separate files and then do some image magic to combine them.

**Comment by zilav** [comment]
They are separate images in script, you'll have to combine them manually.

**Comment by AndalayBay** [comment]
Yes, I see that. Is there a way to combine them in the script? They are also different sizes, so that would have to be dealt with as well.

**Comment by zilav** [comment]
Map image and overlay have the same size, you probably saved them with different scaling.
It is most likely possible to combine in script, but I don't have time for that right now.

**Comment by AndalayBay** [comment]
No, I just confirmed it. I modified the script so that it opens the save dialogue twice. The first time is for the map and the second is for the overlay. I give the files different names. The map is 4096x4096 and the overlay is 1024x1024. I found the code where the overlay is resized to match the map, but that doesn't seem to carry over to the save.

Anyway this suits my purposes. Thanks again for the help.
